### DISCUSSION TITLE
Deno support?

### DISCUSSION THREAD

**Discussion by lucsoft** [panva/node-oidc-provider#1250] [Ideas]
Great thats its ESM Only, would now be nice if i could use Deno.serve instead instead of the current http server

**Comment by panva** [comment]
Sorry, Node.js runtimes only.

**Reply by lucsoft** [reply]
btw why? NodeJS now support Request and Response which would be a super abstraction around handling HTTP so maybe not giving first class support but at least allows to hook into it would be nice

**Reply by panva** [reply]
> why?

I have no interest in running `oidc-provider` outside of the Node.js runtime. That's not to say I will never have such interest, but if that was the case I'd rather start a fresh WebAPI based project than undergo a massive refactor of this one.

> NodeJS now support Request and Response

There's no translation from Node's native `http.IncomingMessage` and `http.ServerResponse` to `Request` and `Response` in Node.js at the moment, and none is planned. Request/Response in Node.js is an artifact of the fetch() API, nothing more.

But if {insert runtime} marketing messages are anything to go by the {insert runtime} Node compatibility layer should allow you to run a `node:http` server and execute `oidc-provider` `Provider.prototype.callback` in it. That of course would require the other `node:*` compat modules in {insert runtime} to actually be compatible, which, let's face it, they aren't.

**Reply by ImSingee** [reply]
+1 for different runtimes (deno, bun, cloudflare workers, etc.)

I think the main problem for compatibility with other runtimes is that node-oidc-provider is tied to Koa. Maybe we can change the server handler to some other (like hono) in the next major release.

**Reply by panva** [reply]
I appreciate the interest.

However, no need to + or - here. It is not in the cards/stars at all.

As a side note, oidc-provider is not tied to koa, it uses koa, yes, but what it is tied to is node's `http.IncomingMessage` and `http.ServerResponse` which is then abstracted over with koa.

If you want to use this in {X} then I'd suggest to go to {X} issue tracker and inquire about the progress of their promised node compatibility.

**Reply by EtienneK** [reply]
Subscribing to this in hopes something changes.

Been asking for this for  years: https://twitter.com/EtienneK/status/1541821393360326657

**Comment by EtienneK** [comment]
I got `node-oidc-provider` to work with Deno. 

To get around the WebAPI `Request`/`Response` translation to/from `IncomingMessage`/`ServerResponse`, I basically built some wrappers that would do the work for me. 

See `NodeRequest` and  `NodeResponse`, here: <https://github.com/EtienneK/efauth/blob/main/oidc/oidc.ts>
and their usage here: <https://github.com/EtienneK/efauth/blob/main/routes/oidc/%5B...oidc%5D.ts> and here: <https://github.com/EtienneK/efauth/blob/main/routes/interactions/%5Buid%5D/login.tsx>

Not pretty, but it works.

Check out a demo here (running on Deno Deploy): <https://efauth.deno.dev>
